File: ml-sidecar/features.py
"""
Point-in-time technical-analysis features — a faithful port of the Node
backend's features.js so the sidecar models compete on identical inputs.

Every function only ever inspects data at-or-before the requested index, which is
what keeps the walk-forward backtest leak-free.
"""
from __future__ import annotations
import math


def log_returns(closes):
    out = []
    for i in range(1, len(closes)):
        a, b = closes[i - 1], closes[i]
        if a > 0 and b > 0:
            out.append(math.log(b / a))
    return out


def _mean(a):
    return sum(a) / len(a) if a else 0.0


def _stdev(a):
    if len(a) < 2:
        return 0.0
    m = _mean(a)
    return math.sqrt(sum((x - m) ** 2 for x in a) / (len(a) - 1))


def _sma(arr, n):
    if len(arr) < n:
        return None
    return sum(arr[-n:]) / n


def _rsi(prices, period=14):
    if len(prices) < period + 1:
        return None
    gains = losses = 0.0
    for i in range(len(prices) - period, len(prices)):
        diff = prices[i] - prices[i - 1]
        if diff >= 0:
            gains += diff
        else:
            losses -= diff
    if losses == 0:
        return 100.0
    rs = (gains / period) / (losses / period)
    return 100.0 - 100.0 / (1.0 + rs)


def daily_vol(closes, window=30):
    r = log_returns(closes[-(window + 1):])
    return _stdev(r) if len(r) >= 2 else 0.0


def feature_row(closes, j):
    """8-feature vector for index j (uses closes[0..j] only), or None."""
    if j < 50:
        return None
    win = closes[max(0, j - 59): j + 1]   # ≤50-day lookback; matches features.js
    price = win[-1]
    if not price > 0:
        return None
    r = log_returns(win)
    ret1 = r[-1] if r else 0.0
    ret5 = _mean(r[-5:]) if len(r) >= 5 else 0.0
    ret10 = _mean(r[-10:]) if len(r) >= 10 else 0.0
    s10, s20, s50 = _sma(win, 10), _sma(win, 20), _sma(win, 50)
    rsi14 = _rsi(win, 14)
    vol30 = daily_vol(win, 30)
    if s10 is None or s20 is None or s50 is None or rsi14 is None:
        return None
    return [
        ret1, ret5, ret10,
        s10 / price - 1, s20 / price - 1, s50 / price - 1,
        rsi14 / 100 - 0.5, vol30,
    ]


FEATURE_NAMES = ['ret1', 'ret5', 'ret10', 'sma10_gap', 'sma20_gap', 'sma50_gap', 'rsi14_c', 'vol30']

File: backend/src/services/prediction/features.js
/**
 * Point-in-time feature helpers for the prediction competition.
 *
 * Everything here takes a chronological `closes` array (oldest -> newest) and an
 * index, and only ever looks at data at-or-before that index. This is what keeps
 * the walk-forward backtest leak-free: a model evaluated "as of" index i is handed
 * closes.slice(0, i+1) and can never see the future.
 *
 * Reuses the technical-indicator math from ../prediction.js so there is a single
 * source of truth.
 */
import { sma, rsi, stdev } from '../prediction.js';

export function mean(arr) {
  if (!arr.length) return 0;
  return arr.reduce((a, b) => a + b, 0) / arr.length;
}

/** Daily log returns of a closes series: r_t = ln(c_t / c_{t-1}). */
export function logReturns(closes) {
  const r = [];
  for (let i = 1; i < closes.length; i++) {
    const a = closes[i - 1], b = closes[i];
    if (a > 0 && b > 0) r.push(Math.log(b / a));
  }
  return r;
}

/** Annualization-free daily volatility = stdev of trailing `window` log returns. */
export function dailyVol(closes, window = 30) {
  const r = logReturns(closes.slice(-(window + 1)));
  return r.length >= 2 ? stdev(r) : 0;
}

/**
 * Feature row for the index `j` of `closes` (uses closes[0..j] only).
 * Returns a fixed-length numeric vector, or null if there isn't enough history.
 * Used both to build a model's training matrix and to form the prediction input.
 */
export function featureRow(closes, j) {
  if (j < 50) return null;                 // need SMA50 + RSI windows
  // All features look back ≤ 50 days, so bound the slice to the last 60 closes
  // (still strictly data at-or-before j — keeps it leak-free and O(1) per call,
  // which matters because the ridge model builds a training matrix from this).
  const win = closes.slice(Math.max(0, j - 59), j + 1);
  const price = win[win.length - 1];
  if (!(price > 0)) return null;
  const r = logReturns(win);
  const ret1 = r.at(-1) ?? 0;
  const ret5 = r.length >= 5 ? mean(r.slice(-5)) : 0;
  const ret10 = r.length >= 10 ? mean(r.slice(-10)) : 0;
  const s10 = sma(win, 10), s20 = sma(win, 20), s50 = sma(win, 50);
  const rsi14 = rsi(win, 14);
  const vol30 = dailyVol(win, 30);
  if (s10 == null || s20 == null || s50 == null || rsi14 == null) return null;
  return [
    ret1,
    ret5,
    ret10,
    s10 / price - 1,
    s20 / price - 1,
    s50 / price - 1,
    rsi14 / 100 - 0.5,
    vol30,
  ];
}

export const FEATURE_NAMES = [
  'ret1', 'ret5', 'ret10', 'sma10_gap', 'sma20_gap', 'sma50_gap', 'rsi14_c', 'vol30',
];
